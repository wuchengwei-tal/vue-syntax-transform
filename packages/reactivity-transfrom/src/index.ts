// API
export { parse } from "./parse";
export { compileTemplate } from "./compileTemplate";
export { compileScript } from "./compileScript";
export { rewriteDefault } from "./rewriteDefault";
export {
    shouldTransform as shouldTransformRef,
    transform as transformRef,
    transformAST as transformRefAST,
} from "@vue/reactivity-transform";

// Utilities
export { parse as babelParse } from "@babel/parser";
import MagicString from "magic-string";
export { MagicString };
// technically internal but we want it in @vue/repl, cast it as any to avoid
// relying on estree types
import { walk as _walk } from "estree-walker";
export const walk = _walk as any;

// Types
export {
    SFCParseOptions,
    SFCParseResult,
    SFCDescriptor,
    SFCBlock,
    SFCTemplateBlock,
    SFCScriptBlock,
    SFCStyleBlock,
} from "./parse";
export {
    TemplateCompiler,
    SFCTemplateCompileOptions,
    SFCTemplateCompileResults,
} from "./compileTemplate";

export { SFCScriptCompileOptions } from "./compileScript";
export {
    AssetURLOptions,
    AssetURLTagConfig,
} from "./templateTransformAssetUrl";
