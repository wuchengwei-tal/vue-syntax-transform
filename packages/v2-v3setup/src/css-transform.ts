import type { CssNode, Dimension, Raw, Selector } from 'css-tree'
import MagicString from 'magic-string'
const csstree = require('css-tree')

const Ratio = 100

export function cssTransform(css: string) {
  return _cssTransform(transComment(css))
}

export function _cssTransform(css: string, raw = '') {
  const s = new MagicString(css)
  const ast = csstree.parse(css, { positions: true })

  function transDimension(node: Dimension) {
    if (node.unit === 'rem') {
      let { start, end } = node.loc!
      const val = parseFloat(node.value) * Ratio + 'px'
      s.overwrite(start.offset, end.offset, val)
    }
  }

  function transSelector(node: Selector) {
    const head = node.children.first
    if (head?.type === 'Combinator' && /\/\s*deep\s*\//.test(head?.name)) {
      const { start, end } = head.loc!
      s.overwrite(start.offset, end.offset, ':deep(')
      s.appendRight(node.loc!.end.offset, ')')
    }
  }

  function transRaw(node: Raw) {
    if (node.value !== raw) {
      const { start, end } = node.loc!
      const css = _cssTransform(node.value, node.value)
      s.overwrite(start.offset, end.offset, css)
    }
  }

  csstree.walk(ast, {
    leave(node: CssNode) {
      if (node.type === 'Dimension') transDimension(node)

      if (node.type === 'Selector') transSelector(node)

      if (node.type === 'Raw') transRaw(node)
    }
  })

  return s.toString()
}

export function transComment(css: string) {
  let code = ''
  for (let line of css.split('\n')) {
    if (/\s*\/\/.*/.test(line)) {
      line = line.replace('//', '/*')
      line += ' */'
    }

    code += line + '\n'
  }
  return code
}
