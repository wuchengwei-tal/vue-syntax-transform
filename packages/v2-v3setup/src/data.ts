export enum BindingTypes {
  /**
   * returned from data()
   */
  DATA = 'data',
  /**
   * declared as a prop
   */
  PROPS = 'props',
  /**
   * a let binding
   */
  LET = 'let',
  /**
   * a const binding that.
   */
  CONST = 'const',

  COMPUTED = 'computed',

  FILTER = 'filter',

  METHOD = 'method',

  WATCH = 'watch',

  HOOK = 'hook',

  $ = '$'
}

export type BindingMap<T> = Record<string, { type: BindingTypes; value: T }>

export const LifeCircleHookMap = {
  beforeMount: 'onBeforeMount',
  mounted: 'onMounted',
  beforeUpdate: 'onBeforeUpdate',
  updated: 'onUpdated',
  activated: 'onActivated',
  deactivated: 'onDeactivated',
  beforeDestroy: 'onBeforeUnmount',
  destroyed: 'onUnmounted',
  beforeRouteEnter: 'onBeforeRouteUpdate',
  beforeRouteLeave: 'onBeforeRouteLeave',
  beforeCreate: '',
  created: ''
}

export const WalkOrder = [
  'data',
  'methods',
  'computed',
  'filter',
  'watch',
  ...Object.keys(LifeCircleHookMap)
]

export const OutputOrder = [
  BindingTypes.CONST,
  BindingTypes.LET,
  BindingTypes.PROPS,
  BindingTypes.DATA,
  BindingTypes.COMPUTED,
  BindingTypes.FILTER,
  BindingTypes.METHOD
]

export const RenderFunction = 'RenderFunction'
