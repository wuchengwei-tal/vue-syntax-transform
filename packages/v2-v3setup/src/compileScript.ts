import { parse as _parse, ParserOptions } from '@babel/parser'

import {
  Node,
  ObjectExpression,
  ObjectMethod,
  ObjectProperty,
  Program,
  BlockStatement,
  Identifier,
  StringLiteral,
  Expression,
  Declaration,
  ArrayPattern,
  ObjectPattern,
  identifier
} from '@babel/types'

import MagicString from 'magic-string'

import { generateCodeFrame } from '@vue/shared'

import { BindingTypes } from '@vue-transform/shared'
import { BindingMap, RenderFunction, VModel } from './data'
import { transformBindings, registerBinding } from './transform'

export function compileScript(
  source: string,
  options: { id: string }
): { content: string; bindings?: Record<string, BindingTypes> } {
  if (!source) return { content: source }

  // metadata that needs to be returned
  const bindingMetadata: Record<string, BindingTypes> = {}
  const optionsBindings: BindingMap<any> = Object.create(null)
  const watcherBindings: BindingMap<ObjectProperty['value'] | ObjectMethod> =
    Object.create(null)
  const hookBindings: BindingMap<ObjectMethod> = Object.create(null)
  const model: VModel = { exist: false, prop: 'value', event: 'input' }

  let name = ''
  const mixins = []
  const assets: string[] = []
  const components: { alias?: string; name: string }[] = []
  let defaultExport: Node | undefined
  const s = new MagicString(source)
  const startOffset = 0
  const endOffset = source.length

  function parse(
    input: string,
    _options: ParserOptions,
    offset: number
  ): Program {
    try {
      return _parse(input, _options).program
    } catch (e: any) {
      e.message = `[@vue/compiler-sfc] ${e.message}\n\n${
        options.id
      }\n${generateCodeFrame(source, e.pos + offset, e.pos + offset + 1)}`
      throw e
    }
  }

  function walkOptions(node: ObjectExpression) {
    for (const property of node.properties) {
      if (property.type !== 'SpreadElement') {
        walkOption(property)
      } else {
        if (property.argument.type === 'ObjectExpression')
          walkOptions(property.argument)
      }
    }
  }

  function walkOption(property: ObjectProperty | ObjectMethod) {
    const { key } = property
    let name = ''
    if (key.type === 'Identifier') name = key.name
    if (key.type === 'StringLiteral') name = key.value
    if (property.type === 'ObjectMethod') {
      if (name === 'data') walkState(optionsBindings, property.body, assets)
      else if (name === 'render')
        registerBinding(
          optionsBindings,
          identifier(RenderFunction),
          property,
          BindingTypes.METHOD
        )
      else {
        registerBinding(
          hookBindings,
          identifier(name),
          property,
          BindingTypes.HOOK
        )
      }
    }

    if (property.type === 'ObjectProperty') {
      if (name === 'name' && property.value.type === 'StringLiteral')
        name = property.value.value

      if (name === 'components') {
        const { value } = property
        if (value.type === 'ObjectExpression') {
          for (const component of value.properties) {
            if (component.type === 'ObjectProperty') {
              const { key, value } = component
              if (key.type === 'Identifier' && value.type === 'Identifier') {
                const { name } = value
                const alias = key.name
                alias === name
                  ? components.push({ name })
                  : components.push({ alias, name })
              }
            }
          }
        }
      }

      if (name === 'mixins') {
        const { value } = property
        if (value.type === 'ArrayExpression') {
          for (const mixin of value.elements) {
            if (mixin?.type === 'Identifier') mixins.push(mixin.name)
          }
        }
      }

      if (name === 'props') walkProps(optionsBindings, property)

      if (name === 'model') updateModel(model, property)

      if (name === 'filters')
        walkMethods(optionsBindings, property, BindingTypes.FILTER)

      if (name === 'computed') walkGetter(optionsBindings, property)

      if (name === 'methods')
        walkMethods(optionsBindings, property, BindingTypes.METHOD)

      if (name === 'watch') walkWatches(watcherBindings, property)
    }
  }

  function walkProps(bindings: BindingMap<any>, property: ObjectProperty) {
    const { value } = property
    if (value.type === 'ObjectExpression') {
      if (value.properties.length) {
        const props = property.key as Identifier
        registerBinding(bindings, props, value, BindingTypes.$)
      }
      for (const p of value.properties) {
        if (p.type === 'ObjectProperty') {
          const { key, value } = p
          if (key.type === 'Identifier') {
            if (key.name === model.prop) {
              if (!model.exist) {
                model.exist = /emit\(('|")input/.test(source)
              }
              registerBinding(
                bindings,
                identifier(key.name),
                value,
                BindingTypes.PROPS
              )
              key.name = 'modelValue'
            } else registerBinding(bindings, key, null, BindingTypes.PROPS)
          }
        }
      }
    }
  }

  // 0. parse both <script> blocks
  const scriptAst = parse(
    source,
    { plugins: [], sourceType: 'module' },
    startOffset
  )
  if (!scriptAst) return { content: source }

  for (const node of scriptAst.body) {
    if (node.type === 'ImportDeclaration') {
      const { source, specifiers } = node
      for (const specifier of specifiers) {
        if (specifier.type === 'ImportDefaultSpecifier') {
          if (!/\.(vue|js)/.test(source.value) && source.value.includes('.'))
            assets.push(specifier.local.name)
        }
      }
    }

    if (node.type === 'ExportNamedDeclaration') {
      s.remove(startOffset + node.start!, startOffset + node.end!)
    }

    if (node.type === 'VariableDeclaration') {
      walkDeclaration(node, optionsBindings)
    }

    if (node.type === 'ExportDefaultDeclaration') {
      const { declaration } = node

      if (declaration.type === 'ObjectExpression') {
        walkOptions(declaration)
      }

      // export default
      defaultExport = node
    }
  }

  // . remove non-script content
  s.remove(0, startOffset)
  s.remove(endOffset, source.length)

  for (const key in optionsBindings) {
    bindingMetadata[key] = optionsBindings[key].type
  }

  const code = transformBindings(
    optionsBindings,
    watcherBindings,
    hookBindings,
    model
  )

  if (defaultExport) {
    s.prependRight(endOffset, code + '\n')
    s.remove(
      startOffset + defaultExport.start!,
      startOffset + defaultExport.end!
    )
  }

  s.trim()

  return {
    bindings: bindingMetadata,
    content: s.toString()
  }
}

function walkState(
  bindings: BindingMap<ObjectProperty['value']>,
  block: BlockStatement,
  assets: string[]
) {
  for (const node of block.body) {
    if (node.type === 'ReturnStatement') {
      const { argument } = node
      if (argument?.type === 'ObjectExpression') {
        for (const state of argument.properties) {
          if (state.type === 'ObjectProperty') {
            const { key, value } = state
            if (key.type === 'Identifier' && !assets.includes(key.name))
              registerBinding(bindings, key, value, BindingTypes.DATA)
          }
        }
      }
    }
  }
}

function walkGetter(
  bindings: BindingMap<BlockStatement | string | Expression>,
  computed: ObjectProperty
) {
  if (computed.value.type === 'ObjectExpression') {
    for (const property of computed.value.properties) {
      if (property.type === 'ObjectMethod') {
        const { key, body } = property
        if (key.type === 'Identifier')
          registerBinding(bindings, key, body, BindingTypes.COMPUTED)
      }

      if (property.type === 'ObjectProperty') {
        const { key, value } = property
        if (key.type === 'Identifier') {
          if (
            value.type === 'FunctionExpression' ||
            value.type === 'ArrowFunctionExpression'
          ) {
            registerBinding(bindings, key, value.body, BindingTypes.COMPUTED)
          }
        }
      }

      if (property.type === 'SpreadElement') {
        const { argument } = property
        if (argument.type === 'CallExpression') {
          const { callee } = argument
          if (callee.type === 'Identifier' && callee.name === 'mapGetters') {
            const items = argument.arguments
            if (items.length && items[0].type === 'ObjectExpression') {
              for (const item of items[0].properties) {
                if (item.type !== 'ObjectProperty') continue
                const { key, value } = item
                const val = (value as StringLiteral).value
                if (key.type === 'Identifier')
                  registerBinding(bindings, key, val, BindingTypes.COMPUTED)
              }
            }
          }
        }
      }
    }
  }
}

function walkMethods(
  bindings: BindingMap<ObjectMethod>,
  property: ObjectProperty,
  type: BindingTypes
) {
  const { value } = property
  if (value.type === 'ObjectExpression') {
    for (const func of value.properties) {
      if (func.type === 'ObjectMethod') {
        if (func.key.type === 'Identifier')
          registerBinding(bindings, func.key, func, type)
      }
    }
  }
}

function walkWatches(
  bindings: BindingMap<ObjectProperty['value'] | ObjectMethod>,
  property: ObjectProperty
) {
  const { value } = property
  if (value.type === 'ObjectExpression') {
    for (const watcher of value.properties) {
      if (watcher.type === 'ObjectProperty') {
        const { key, value } = watcher
        if (key.type === 'Identifier')
          registerBinding(bindings, key, value, BindingTypes.WATCH)

        if (key.type === 'StringLiteral') {
          registerBinding(
            bindings,
            identifier(key.value),
            value,
            BindingTypes.WATCH
          )
        }
      }
      if (watcher.type === 'ObjectMethod') {
        if (watcher.key.type === 'Identifier')
          registerBinding(bindings, watcher.key, watcher, BindingTypes.WATCH)
      }
    }
  }
}

function walkDeclaration(node: Declaration, bindings: BindingMap<any>) {
  if (node.type === 'VariableDeclaration') {
    const isConst = node.kind === 'const'
    const bindingType = isConst ? BindingTypes.CONST : BindingTypes.LET
    for (const { id, init } of node.declarations) {
      if (id.type === 'Identifier') {
        registerBinding(bindings, id, init, bindingType)
      }
      if (id.type === 'ObjectPattern') {
        walkObjectPattern(id, bindings, isConst)
      } else if (id.type === 'ArrayPattern') {
        walkArrayPattern(id, bindings, isConst)
      }
    }
  } else if (
    node.type === 'FunctionDeclaration' ||
    node.type === 'ClassDeclaration'
  ) {
    // export function foo() {} / export class Foo {}
    // export declarations must be named.
    registerBinding(bindings, node.id!, node.body, BindingTypes.CONST)
  }
}

function walkObjectPattern(
  node: ObjectPattern,
  bindings: BindingMap<any>,
  isConst: boolean
) {
  for (const p of node.properties) {
    if (p.type === 'ObjectProperty') {
      if (p.key.type === 'Identifier' && p.key === p.value) {
        // shorthand: const { x } = ...
        const type = isConst ? BindingTypes.CONST : BindingTypes.LET
        registerBinding(bindings, p.key, p.value, type)
      } else {
        walkPattern(p.value, bindings, isConst)
      }
    } else {
      // ...rest
      // argument can only be identifier when destructuring
      const type = isConst ? BindingTypes.CONST : BindingTypes.LET
      registerBinding(bindings, p.argument as Identifier, '', type)
    }
  }
}

function walkArrayPattern(
  node: ArrayPattern,
  bindings: BindingMap<any>,
  isConst: boolean
) {
  for (const e of node.elements) {
    e && walkPattern(e, bindings, isConst)
  }
}

function walkPattern(node: Node, bindings: BindingMap<any>, isConst: boolean) {
  if (node.type === 'Identifier') {
    const type = isConst ? BindingTypes.CONST : BindingTypes.LET
    registerBinding(bindings, node, '', type)
  } else if (node.type === 'RestElement') {
    // argument can only be identifier when destructuring
    const type = isConst ? BindingTypes.CONST : BindingTypes.LET
    registerBinding(bindings, node.argument as Identifier, '', type)
  } else if (node.type === 'ObjectPattern') {
    walkObjectPattern(node, bindings, isConst)
  } else if (node.type === 'ArrayPattern') {
    walkArrayPattern(node, bindings, isConst)
  } else if (node.type === 'AssignmentPattern') {
    if (node.left.type === 'Identifier') {
      const type = isConst ? BindingTypes.CONST : BindingTypes.LET
      registerBinding(bindings, node.left, '', type)
    } else {
      walkPattern(node.left, bindings, isConst)
    }
  }
}

function updateModel(model: VModel, property: ObjectProperty) {
  model.exist = true
  const { value } = property
  if (value.type === 'ObjectExpression') {
    for (const p of value.properties) {
      if (p.type === 'ObjectProperty') {
        const { key, value } = p
        if (key.type === 'Identifier' && value.type === 'StringLiteral') {
          if (key.name === 'prop') model.prop = value.value
          if (key.name === 'event') model.event = value.value
        }
      }
    }
  }
}
