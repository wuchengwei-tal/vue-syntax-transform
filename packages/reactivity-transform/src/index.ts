export { compileScript } from './compileScript'

import { compileScript } from './compileScript'

export { SFCScriptCompileOptions } from './compileScript'

import { parse } from '@vue/compiler-sfc'

export function reactivityTransform(src: string, id = 'xxxxxx') {
  const { descriptor } = parse(src)
  return compileScript(descriptor, {
    id,
    reactivityTransform: true
  })
}
