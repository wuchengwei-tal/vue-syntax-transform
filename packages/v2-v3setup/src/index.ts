import {
  type SFCDescriptor,
  parseComponent,
  compile
} from 'vue-template-compiler'

import { compileScript } from './compileScript'

import { templateTransform } from './template-transform'
import { cssTransform } from './css-transform'
import MagicString from 'magic-string'

export * from './data'

type TransformedSFC = {
  template?: ReturnType<typeof templateTransform>
  script?: ReturnType<typeof compileScript>
  styles?: ReturnType<typeof cssTransform>[]
  content: string
}

export function sfcTransform(sfc: string, id = '') {
  const descriptor = parseComponent(sfc) as SFCDescriptor & {
    source: string
    cssVars: string[]
    errors: any[]
  }

  const { script, template, styles, source } = descriptor

  const s = new MagicString(source)

  const ret: TransformedSFC = { content: '', styles: [] }

  let inheritAttrs = true
  if (script) {
    // console.log(descriptor)
    const match = script.content.match(/inheritAttrs\s*:\s*(true|false)/)
    if (match) inheritAttrs = match[1] === 'true'
  }

  if (template) {
    const { content, start, end } = template

    const { ast } = compile(content, {
      outputSourceRange: true,
      whitespace: 'preserve'
    })
    ret.template = templateTransform({ ast, ...template, inheritAttrs })
    s.overwrite(start!, end!, ret.template.content)
  }

  if (script) {
    const { start, end, content } = script
    if (content) {
      ret.script = compileScript(content, { id })
      s.overwrite(start!, end!, '\n' + ret.script.content + '\n')
      let i = start!
      while (source[i--] === '>');
      while (source[--i] !== '<');
      if (source.slice(i, i + 7) === '<script') {
        s.prependRight(i + 7, ' setup')
      }
    }
  }

  if (styles.length) {
    styles.forEach(style => {
      const { start, end, content } = style
      if (content) {
        const res = cssTransform(content)
        ret.styles?.push(res)
        s.overwrite(start!, end!, res)
      }
    })
  }

  return { ...ret, content: s.toString() }
}

function transformActions(code: string) {
  const fnRe = /(export )?const \w+ = (async )?\(.*[\)|\n]/g

  let match
  let last = 0
  let output = ''
  while ((match = fnRe.exec(code))) {
    const pre = code.slice(last, match.index)

    let [line, _, __] = match

    last = match.index + line.length

    let flag = line[line.length - 2]
    // console.log(line);
    if (line) {
      // oneline params
      if (flag === '{') {
        const match = /\(((\{.*\})|(\w+)),? ?(.*)?\)/.exec(line)
        if (match) {
          let [arg1, _, __, ___, arg2e] = match
          line = line.replace(arg1.slice(1, -1), arg2e || '')
        }
      }
      output += pre + line
      // multi line params
      if (flag === '(') {
        const params = parseInside(code, last, '(')
        let i = 0
        while (params[i++] !== '}');
        if (params[i] === ',') i++
        output += params.slice(i)

        last += params.length
        while (code[last++] !== '{') output += code[last]
        flag = '{'
      }
    }

    if (flag === '{') {
      let body = parseInside(code, last, '{')
      last += body.length
      body = replaceActionBody(body)
      body = replaceState(body)
      output += body
    }
  }
  output += code.slice(last)

  // console.log(output);
  return output
}

const Pair = { '{': '}', '(': ')', '[': ']' }
function parseInside(code: string, s: number, p: keyof typeof Pair) {
  let e = s
  let cnt = 0
  while (e < code.length - 1) {
    const ch = code[e++]
    if (ch === p) cnt++
    if (ch === Pair[p]) {
      if (cnt) cnt--
      else break
    }
  }
  const body = code.slice(s, e)
  return body
}

function replaceActionBody(code: string) {
  const dispatchRe = /(((store\.)?dispatch)|(commit))\('?((\w+\.?\w+))'?,? ?/g

  let match
  let last = 0
  let output = ''
  while ((match = dispatchRe.exec(code))) {
    const pre = code.slice(last, match.index)

    let [line, _, store, dispatch, commit, fn] = match

    last = match.index + line.length

    if (line && fn) {
      if (fn.includes('.')) fn = fn.split('.')[1]
      line = fn + '('
      // console.log(line)
    }

    output += pre + line
  }
  output += code.slice(last)
  // console.log(output);
  return output
}

function transformMutations(code: string) {
  const { body, index } = getExportDefault(code)
  if (!body) return code

  let last = index
  let output = code.slice(0, index)

  let match
  const mutationRe = /((\w+)|\[types.(\w+)\])(\(state,?.*\)) {/g
  while ((match = mutationRe.exec(body))) {
    let pre = ''
    if (last !== index) pre = body.slice(last, match.index)

    let [line, _, name, type, params] = match

    last = match.index + line.length
    params = params.replace(/state,? ?/, '')
    line = 'export function ' + (name || type) + params + '{'

    let mutationBody = parseInside(body, last, '{')
    last += mutationBody.length + 1 // del comma

    mutationBody = replaceState(mutationBody)

    output += pre + line + mutationBody
  }

  return output
}

function getExportDefault(code: string) {
  const defaultRe = /export default {/
  const match = defaultRe.exec(code)
  if (match) {
    return {
      body: parseInside(code, match.index, '{'),
      index: match.index
    }
  }
  return {}
}

function replaceState(code: string) {
  const stateRe = /(state\.\w+)( |\(|\.|\[|,)/g
  let last = 0
  let output = ''
  let match
  while ((match = stateRe.exec(code))) {
    const pre = code.slice(last, match.index)

    let [res, state] = match
    last = match.index + res.length

    if (res.endsWith('.')) res = res.replace(res, `${res}value.`)
    else res = res.replace(state, `${state}.value`)
    res = res.replace('state.', '')

    output += pre + res
  }
  output += code.slice(last)

  output = output.replace(/rootState\??\.\w+\??\./g, '')

  return output
}
