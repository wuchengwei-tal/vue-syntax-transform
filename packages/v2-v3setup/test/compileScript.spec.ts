import { expect } from 'vitest'

import { BindingTypes, sfcTransform } from '../src'
import { assertCode } from './utils'

describe('script transform', () => {
  test('data', () => {
    const { content, bindings } = sfcTransform(`
      <script>
      import png from './a.png'
      export default {
        data() {
          return {
            firstName: 'Foo',
            lastName: 'Bar',
            n: 1,
            bool: true,
            obj: {}, 
            timeInterval: null,
            list: [],
            png,
          };
        }
      }
      </script>
      `).script!

    expect(bindings).toStrictEqual({
      firstName: BindingTypes.DATA,
      lastName: BindingTypes.DATA,
      n: BindingTypes.DATA,
      bool: BindingTypes.DATA,
      obj: BindingTypes.DATA,
      timeInterval: BindingTypes.DATA,
      list: BindingTypes.DATA
    })

    expect(content).not.toMatch(`const png = ref`)
    assertCode(content)
  })

  test('computed', () => {
    const { content, bindings } = sfcTransform(`
    <script>
    export default {
      data() {
        return {
          firstName: 'Foo',
          lastName: 'Bar',
        };
      },
      computed: {
        fullName: function () {
          return this.firstName + ' ' + this.lastName
        }
      },
    }
    </script>
    `).script!

    expect(bindings).toStrictEqual({
      firstName: BindingTypes.DATA,
      lastName: BindingTypes.DATA,
      fullName: BindingTypes.COMPUTED
    })

    expect(content).toMatch('const firstName = ref(')
    expect(content).toMatch('const fullName = computed(() => {')

    assertCode(content)
  })

  test('watch', () => {
    const { content, bindings } = sfcTransform(`
    <script>
    export default {
      data() {
        return {
          firstName: 'Foo',
          lastName: 'Bar',
          fullName: {
            a: 1
          },
        };
      },
      watch: {
        firstName: 'onFirstName',
        lastName: {
          handler(v1, v2){
            if(v1 !== v2) {
              this.firstName = 'firstName'
            }
          },
          deep: true,
        },
        fullName(v1, v2) {
          this.firstName = 'fullName'
        },
        'fullName.a' : {
          handler(v1, v2){
          },
        }
      }
    }
    </script>
    `).script!

    expect(bindings).toStrictEqual({
      firstName: BindingTypes.DATA,
      lastName: BindingTypes.DATA,
      fullName: BindingTypes.DATA
    })

    expect(content).toMatch('watch(firstName, onFirstName')
    expect(content).toMatch('watch(lastName, (v1, v2) => {')
    expect(content).toMatch('watch(fullName, (v1, v2) => {')
    expect(content).toMatch('watch(() => fullName.a, (v1, v2)')

    assertCode(content)
  })

  test('$refs', () => {
    const { content } = sfcTransform(`
    <script>
      export default {
        mounted(){
          this.$refs?.listRef.page
        }
      }
    </script>
    `).script!
    expect(content).toMatch('const listRef = ref()')
    expect(content).toMatch(' listRef.page')
    expect(content).not.toMatch('mounted(')
    expect(content).toMatch('onMounted(()')

    assertCode(content)
  })

  test('router', () => {
    const { content } = sfcTransform(`
    <script>
      export default {
        created(){
          this.$route.fullPath
          this.$router.push('/list')
        }
      }
    </script>
    `).script!
    expect(content).toMatch('const route = useRoute()')
    expect(content).toMatch(' route.fullPath')
    expect(content).toMatch('const router = useRouter()')
    expect(content).toMatch(" router.push('/list');")
    expect(content).not.toMatch('created(')

    assertCode(content)
  })

  test('emit', () => {
    const { content } = sfcTransform(`
    <script>
      export default {
        created(){
          this.$emit('click', 0)
          this.$emit('change', 1)
        },
      }
    </script>
    `).script!

    expect(content).toMatch(`const emit = defineEmits(["click", "change"])`)
    expect(content).toMatch(` emit('click', 0)`)
    expect(content).toMatch(` emit('change', 1)`)

    assertCode(content)
  })

  test('vuex', () => {
    const { content } = sfcTransform(`
    <script>
      export default {
        created(){
          this.$store.commit('setLoadingState', false);
          this.$store.dispatch('getServerTime');
          this.$store.registerModule('module');
          this.$store.unregisterModule('module');
          const a = this.$store.state.a;
          const b = this.$store.getter.b;
        },
      }
    </script>
    `).script!

    expect(content).not.toMatch(`this.$store`)
    expect(content).not.toMatch(`registerModule`)
    expect(content).not.toMatch(`unregisterModule`)
    expect(content).toMatch(` setLoadingState(false)`)
    expect(content).toMatch(` getServerTime()`)
    expect(content).toMatch(`const a = a`)
    expect(content).toMatch(`const b = b`)

    assertCode(content)
  })

  test('props', () => {
    const { content } = sfcTransform(`
    <script>
      export default {
        props: {
          prop: {
            type: Object,
            default: () => {},
          },
          bool: Boolean,
        },
        mounted(){
          this.prop
        },
        
      }
    </script>
    `).script!

    expect(content).not.toMatch(`this.`)
    expect(content).toMatch(`props.prop`)

    assertCode(content)
  })

  test('this', () => {
    const { content } = sfcTransform(`
    <script>
      export default {
        mounted(){
          let self
          this.a
          const that = this
          self = this
          that.a
          ;(function (){
            this.b
            that.c
            that.a
            that.e
          })
          function func(){
            this.d
            that.e
          }
        },
      }
    </script>
    `).script!
    expect(content).toMatch(`this.b`)
    expect(content).toMatch(`this.d`)
    expect(content).not.toMatch(`this.a`)
    expect(content).not.toMatch(`that.c`)
    expect(content).not.toMatch(`that.e`)
    expect(content).not.toMatch(`const that = this`)
    expect(content).toMatch(`const a = ref()`)
    expect(content).toMatch(`const c = ref()`)
    expect(content).toMatch(`const e = ref()`)
    assertCode(content)
  })

  test('render prop', () => {
    const { content } = sfcTransform(`
    <script>
      export default {
        render(h){
          return h('div')
        },
      }
    </script>
    `).script!

    expect(content).not.toMatch(`render(h) {`)
  })

  test('v-model', () => {
    const { content } = sfcTransform(`
    <script>
      export default {
        props: {
          value: String,
        },
        mounted(){
          this.value
          this.$emit('input', 'foo')
        }
      }
    </script>
    `).script!

    expect(content).not.toMatch(`value`)
    expect(content).not.toMatch(`input`)
    expect(content).toMatch(`modelValue`)
    expect(content).toMatch(`update:modelValue`)
  })

  test('custom v-model', () => {
    const { content } = sfcTransform(`
    <script>
      export default {
        model: {
          prop: 'title',
          event: 'change'
        },
        props: {
          value: String,
          title: String,
        },
        mounted(){
          this.title
          this.$emit('change', 'foo')
        }
      }
    </script>
    `).script!

    expect(content).not.toMatch(`title`)
    expect(content).not.toMatch(`change`)
    expect(content).toMatch(`modelValue`)
    expect(content).toMatch(`update:modelValue`)
  })

  test('filter', () => {
    const { content, bindings } = sfcTransform(`
<script>
  export default {
    filters: {
      currencyUSD(value) {
        return '$' + value
      }
    }
  }
</script>
    `).script!

    expect(bindings).toStrictEqual({
      currencyUSD: BindingTypes.FILTER
    })
    expect(content).toMatch(`function currencyUSD`)
  })
})
