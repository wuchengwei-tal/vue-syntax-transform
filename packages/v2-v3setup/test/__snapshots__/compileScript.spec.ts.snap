// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SFC compile > $refs 1`] = `
"const listRef = ref();
onMounted(() => {
  listRef.page;
});"
`;

exports[`SFC compile > computed 1`] = `
"const fullName = computed(() => {
  return firstName.value + ' ' + lastName.value;
});
const lastName = ref('Bar');
const firstName = ref('Foo');"
`;

exports[`SFC compile > data 1`] = `
"import png from './a.png'
      
      const list = ref([]);
const timeInterval = ref(null);
const obj = ref({});
const bool = ref(true);
const n = ref(1);
const lastName = ref('Bar');
const firstName = ref('Foo');"
`;

exports[`SFC compile > emit 1`] = `
"const emit = defineEmits([\\"click\\", \\"change\\"]);
{
  emit('click', 0);
  emit('change', 1);
}"
`;

exports[`SFC compile > props 1`] = `
"const props = defineProps({
  prop: {
    type: Object,
    default: () => {}
  },
  bool: Boolean
});
onMounted(() => {
  props.prop;
});"
`;

exports[`SFC compile > router 1`] = `
"const router = useRouter();
const route = useRoute();
{
  route.fullPath;
  router.push('/list');
}"
`;

exports[`SFC compile > this 1`] = `
"const e = ref();
const c = ref();
const a = ref();
onMounted(() => {
  let self;
  a.value;
  a.value;
  (function () {
    this.b;
    c.value;
    a.value;
    e.value;
  });
  function func() {
    this.d;
    e.value;
  }
});"
`;

exports[`SFC compile > vuex 1`] = `
"{
  setLoadingState(false);
  getServerTime();
  const a = a;
  const b = b;
}"
`;

exports[`SFC compile > watch 1`] = `
"const fullName = ref({
  a: 1
});
const lastName = ref('Bar');
const firstName = ref('Foo');
watch(firstName, () => \\"firstName\\");
watch(lastName, (v1, v2) => {
  if (v1 !== v2) {
    firstName.value = 'firstName';
  }
}, {
  deep: true
});
watch(fullName, (v1, v2) => {
  firstName.value = 'fullName';
});
watch(() => fullName.a, (v1, v2) => {});"
`;
